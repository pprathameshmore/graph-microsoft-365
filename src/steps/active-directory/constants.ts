import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const steps: Record<string, string> = {
  FETCH_ACCOUNT: 'account',
  FETCH_GROUPS: 'groups',
  FETCH_GROUP_MEMBERS: 'group-members',
  FETCH_USERS: 'users',
};

export const entities: Record<string, StepEntityMetadata> = {
  ACCOUNT: {
    resourceName: '[AD] Account',
    _type: 'microsoft_365_account',
    _class: 'Account',
  },
  GROUP: {
    resourceName: '[AD] Group',
    _type: 'azure_user_group',
    _class: 'UserGroup',
  },
  USER: {
    resourceName: '[AD] User',
    _type: 'azure_user',
    _class: 'User',
  },
  ORGANIZATION: {
    resourceName: '[AD] Organization',
    _type: 'azure_organization',
    _class: 'Organization',
  },
  /**
   * The entity used for members of groups which are not one of the ingested
   * directory objects.
   */
  GROUP_MEMEBER: {
    resourceName: '[AD] Group Member',
    _type: 'azure_group_member',
    _class: 'User',
  },
};

export const relationships: Record<string, StepRelationshipMetadata> = {
  ACCOUNT_HAS_USER: {
    _type: 'microsoft_365_account_has_azure_user',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  ACCOUNT_HAS_GROUP: {
    _type: 'microsoft_365_account_has_azure_group',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP._type,
  },
  GROUP_HAS_MEMBER: {
    _type: 'azure_group_has_member',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP_MEMEBER._type,
  },
  GROUP_HAS_USER: {
    _type: 'azure_group_has_user',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  GROUP_HAS_GROUP: {
    _type: 'azure_group_has_group',
    sourceType: entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: entities.GROUP._type,
  },
};

export const DATA_ACCOUNT_ENTITY = entities.ACCOUNT._type;
