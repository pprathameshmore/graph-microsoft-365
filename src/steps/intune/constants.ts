import {
  generateRelationshipType,
  IntegrationError,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
  FETCH_COMPLIANCE_POLICIES_AND_FINDINGS: 'compliance-policies-and-findings',
  FETCH_DEVICE_CONFIGURATIONS_AND_FINDINGS:
    'device-configurations-and-findings',
  FETCH_MANAGED_APPLICATIONS: 'managed-applications',
  FETCH_DETECTED_APPLICATIONS: 'detected-applications',
};

export type ManagedDeviceType =
  | 'user_endpoint'
  | 'workstation'
  | 'laptop'
  | 'desktop'
  | 'computer'
  | 'server'
  | 'smartphone';
export const managedDeviceTypes: ManagedDeviceType[] = [
  'user_endpoint',
  'workstation',
  'laptop',
  'desktop',
  'computer',
  'server',
  'smartphone',
];
export const INTUNE_HOST_AGENT_KEY_PREFIX = 'intune-host-agent:';

export const entities = {
  MULTI_DEVICE: managedDeviceTypes.map((type) => {
    return {
      resourceName: 'Managed Device',
      _type: type,
      _class: ['Device', 'Host'], // Devices will not have a class of 'Device' if the device is not physical
    };
  }),
  HOST_AGENT: {
    resourceName: 'Intune Host Agent',
    _type: 'intune_host_agent',
    _class: 'HostAgent',
  },
  DEVICE_CONFIGURATION: {
    resourceName: 'Device Configuration',
    _type: 'intune_device_configuration',
    _class: ['Configuration', 'ControlPolicy'] as string[],
  },
  COMPLIANCE_POLICY: {
    resourceName: 'Compliance Policy',
    _type: 'intune_compliance_policy',
    _class: ['Configuration', 'ControlPolicy'] as string[],
  },
  NONCOMPLIANCE_FINDING: {
    resourceName: 'Noncompliance Finding',
    _type: 'intune_noncompliance_finding',
    _class: 'Finding',
  },
  MANAGED_APPLICATION: {
    resourceName: 'Managed Application',
    _type: 'intune_managed_application',
    _class: 'Application',
  },
  DETECTED_APPLICATION: {
    resourceName: 'Detected Application',
    _type: 'intune_detected_application',
    _class: 'Application',
  },
} as const;

function createRelationshipForAllDeviceTypes(relationshipMetadata: {
  sourceType?: string;
  targetType?: string;
  _class: RelationshipClass;
}) {
  const { sourceType, targetType, _class } = relationshipMetadata;
  if ((sourceType && targetType) || (!sourceType && !targetType)) {
    throw new IntegrationError({
      message: `'Invalid use of createRelationshipForAllDeviceTypes, you may have either a sourceType or a targetType'`,
      code: 'RELATIONSHIP_IMPLEMENTATION_ERROR',
    });
  }
  return managedDeviceTypes.map((type) => {
    return {
      _type: generateRelationshipType(
        _class,
        sourceType ?? type,
        targetType ?? type,
      ),
      _class,
      sourceType: sourceType ?? type,
      targetType: targetType ?? type,
    };
  });
}

export const relationships = {
  MULTI_USER_HAS_DEVICE: createRelationshipForAllDeviceTypes({
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
  }),
  MULTI_HOST_AGENT_MANAGES_DEVICE: createRelationshipForAllDeviceTypes({
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.MANAGES,
  }),
  HOST_AGENT_ASSIGNED_DEVICE_CONFIGURATION: {
    _type: 'intune_host_agent_assigned_device_configuration',
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.DEVICE_CONFIGURATION._type,
  },
  HOST_AGENT_ASSIGNED_COMPLIANCE_POLICY: {
    _type: 'intune_host_agent_assigned_compliance_policy',
    sourceType: entities.HOST_AGENT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.COMPLIANCE_POLICY._type,
  },
  DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_device_configuration_identified_noncompliance_finding',
    sourceType: entities.DEVICE_CONFIGURATION._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  COMPLIANCE_POLICY_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_compliance_policy_identified_noncompliance_finding',
    sourceType: entities.COMPLIANCE_POLICY._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  MULTI_DEVICE_HAS_NONCOMPLIANCE_FINDING: createRelationshipForAllDeviceTypes({
    targetType: entities.NONCOMPLIANCE_FINDING._type,
    _class: RelationshipClass.HAS,
  }),
  MULTI_DEVICE_ASSIGNED_MANAGED_APPLICATION: createRelationshipForAllDeviceTypes(
    {
      targetType: entities.MANAGED_APPLICATION._type,
      _class: RelationshipClass.ASSIGNED,
    },
  ),
  MULTI_DEVICE_INSTALLED_DETECTED_APPLICATION: createRelationshipForAllDeviceTypes(
    {
      targetType: entities.DETECTED_APPLICATION._type,
      _class: RelationshipClass.INSTALLED,
    },
  ),
  MANAGED_APPLICATION_MANAGES_DETECTED_APPLICATION: {
    _type: 'intune_managed_application_manages_detected_application',
    sourceType: entities.MANAGED_APPLICATION._type,
    _class: RelationshipClass.MANAGES,
    targetType: entities.DETECTED_APPLICATION._type,
  },
} as const;
