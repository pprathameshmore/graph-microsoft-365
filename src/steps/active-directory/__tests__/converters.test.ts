import { RelationshipDirection } from '@jupiterone/integration-sdk-core';
import { entities } from '../constants';
import {
  createAccountEntityWithOrganization,
  createAccountGroupRelationship,
  createAccountUserRelationship,
  createGroupEntity,
  createGroupMemberRelationship,
  createOrganizationEntity,
  createUserEntity,
} from '../converters';
import exampleAccountEntity from './fixtures/exampleAccountEntity';
import exampleGroup from './fixtures/exampleGroup';
import exampleGroupEntity from './fixtures/exampleGroupEntity';
import exampleGroupMemberGroup from './fixtures/exampleGroupMemberGroup';
import exampleGroupMemberOther from './fixtures/exampleGroupMemberOther';
import exampleGroupMemberUser from './fixtures/exampleGroupMemberUser';
import exampleIntegrationInstance from './fixtures/exampleIntegrationInstance';
import exampleOrganization from './fixtures/exampleOrganization';

describe('createAccountEntityWithOrganization', () => {
  test('properties transferred', () => {
    const exampleIntuneConfig = {
      mobileDeviceManagementAuthority: 'mobileDeviceManagementAuthority',
      subscriptionState: 'subscriptionState',
      intuneAccountID: 'intuneAccountID',
    };
    const accountEntity = createAccountEntityWithOrganization(
      exampleIntegrationInstance,
      exampleOrganization,
      exampleIntuneConfig,
    );
    expect(accountEntity).toMatchSnapshot(
      'createAccountEntityWithOrganization',
    );
  });
});

describe('createGroupEntity', () => {
  test('properties transferred', () => {
    expect(createGroupEntity(exampleGroup)).toEqual({
      _class: ['UserGroup'],
      _key: '89fac263-2430-48fd-9278-dacfdfc89792',
      _type: 'azure_user_group',
      _rawData: [
        {
          name: 'default',
          rawData: exampleGroup,
        },
      ],
      classification: undefined,
      createdOn: 1556042765000,
      deletedOn: undefined,
      description: 'descr',
      name: 'test group',
      displayName: 'test group',
      id: '89fac263-2430-48fd-9278-dacfdfc89792',
      mailEnabled: false,
      mailNickname: '8bb2d1c34',
      renewedOn: 1556042765000,
      securityEnabled: true,
    });
  });
});

describe('createUserEntity', () => {
  test('properties transferred', () => {
    const rawData = {
      businessPhones: ['+1 2223334444'],
      displayName: 'Andrew Kulakov',
      givenName: 'Andrew',
      jobTitle: 'test title',
      mail: undefined,
      mobilePhone: '+1 2223334444',
      officeLocation: 'DBP',
      preferredLanguage: undefined,
      surname: 'Kulakov',
      userPrincipalName:
        'admin_test.dualboot.com#EXT#@admintestdualboot.onmicrosoft.com',
      id: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
    };
    expect(createUserEntity(rawData)).toEqual({
      _class: ['User'],
      _key: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
      _type: 'azure_user',
      _rawData: [
        {
          name: 'default',
          rawData,
        },
      ],
      name: 'Andrew Kulakov',
      displayName: 'Andrew Kulakov',
      givenName: 'Andrew',
      firstName: 'Andrew',
      id: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
      jobTitle: 'test title',
      mobilePhone: '+1 2223334444',
      officeLocation: 'DBP',
      preferredLanguage: undefined,
      surname: 'Kulakov',
      lastName: 'Kulakov',
      username:
        'admin_test.dualboot.com#EXT#@admintestdualboot.onmicrosoft.com',
      userPrincipalName:
        'admin_test.dualboot.com#EXT#@admintestdualboot.onmicrosoft.com',
    });
  });
});

describe('createOrganizationEntity', () => {
  test('properties transferred', () => {
    expect(createOrganizationEntity(exampleOrganization)).toEqual({
      _class: ['Organization'],
      _key: '1111111-1111-1111-1111-111111111111',
      _rawData: [
        {
          name: 'default',
          rawData: exampleOrganization,
        },
      ],
      _type: 'azure_organization',
      id: '1111111-1111-1111-1111-111111111111',
      displayName: 'Default Directory',
      website: 'verifiedDomain.onmicrosoft.com',
      emailDomain: 'verifiedDomain.onmicrosoft.com',
    });
  });
});

describe('createAccountGroupRelationship', () => {
  test('properties transferred', () => {
    expect(
      createAccountGroupRelationship(exampleAccountEntity, {
        _key: '89fac263-2430-48fd-9278-dacfdfc89792',
        _class: entities.GROUP._class,
        _type: entities.GROUP._type,
        id: '89fac263-2430-48fd-9278-dacfdfc89792',
        deletedDateTime: undefined,
        classification: undefined,
        createdDateTime: '2019-04-23T18:06:05Z',
        description: 'descr',
        displayName: 'test group',
        groupTypes: [],
        mail: undefined,
        mailEnabled: false,
        mailNickname: '8bb2d1c34',
        onPremisesLastSyncDateTime: undefined,
        onPremisesSecurityIdentifier: undefined,
        onPremisesSyncEnabled: undefined,
        proxyAddresses: [],
        renewedDateTime: '2019-04-23T18:06:05Z',
        securityEnabled: true,
        visibility: undefined,
        onPremisesProvisioningErrors: [],
      }),
    ).toEqual({
      _class: 'HAS',
      _fromEntityKey: 'microsoft_365_account_id',
      _key: 'microsoft_365_account_id|has|89fac263-2430-48fd-9278-dacfdfc89792',
      _toEntityKey: '89fac263-2430-48fd-9278-dacfdfc89792',
      _type: 'microsoft_365_account_has_azure_group',
      displayName: 'HAS',
    });
  });
});

describe('createAccountUserRelationship', () => {
  test('properties transferred', () => {
    expect(
      createAccountUserRelationship(exampleAccountEntity, {
        _key: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
        _class: entities.USER._class,
        _type: entities.USER._type,
        businessPhones: ['+1 2223334444'],
        displayName: 'Andrew Kulakov',
        givenName: 'Andrew',
        jobTitle: 'test title',
        mail: undefined,
        mobilePhone: '+1 2223334444',
        officeLocation: 'DBP',
        preferredLanguage: undefined,
        surname: 'Kulakov',
        userPrincipalName:
          'admin_test.dualboot.com#EXT#@admintestdualboot.onmicrosoft.com',
        id: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
      }),
    ).toEqual({
      _class: 'HAS',
      _fromEntityKey: 'microsoft_365_account_id',
      _key: 'microsoft_365_account_id|has|abf00eda-02d6-4053-a077-eef036e1a4c8',
      _toEntityKey: 'abf00eda-02d6-4053-a077-eef036e1a4c8',
      _type: 'microsoft_365_account_has_azure_user',
      displayName: 'HAS',
    });
  });
});

describe('createGroupMemberRelationship', () => {
  test('properties transferred for users', () => {
    expect(
      createGroupMemberRelationship(exampleGroupEntity, exampleGroupMemberUser),
    ).toEqual({
      _class: 'HAS',
      _key:
        '89fac263-2430-48fd-9278-dacfdfc89792|has|324e8daa-9c29-42a4-a74b-b9893e6d9750',
      _type: 'azure_group_has_member',
      _mapping: {
        relationshipDirection: RelationshipDirection.FORWARD,
        sourceEntityKey: '89fac263-2430-48fd-9278-dacfdfc89792',
        targetFilterKeys: [['_type', '_key']],
        targetEntity: {
          _key: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
          _type: 'azure_user',
          _class: 'User',
          displayName: 'User Name',
          jobTitle: 'Job Title',
          email: 'user@example.com',
        },
      },
      groupId: '89fac263-2430-48fd-9278-dacfdfc89792',
      memberId: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
      memberType: '#microsoft.graph.user',
      displayName: 'HAS',
    });
  });

  test('properties transferred for groups', () => {
    expect(
      createGroupMemberRelationship(
        exampleGroupEntity,
        exampleGroupMemberGroup,
      ),
    ).toEqual({
      _class: 'HAS',
      _key:
        '89fac263-2430-48fd-9278-dacfdfc89792|has|324e8daa-9c29-42a4-a74b-b9893e6d9750',
      _type: 'azure_group_has_member',
      _mapping: {
        relationshipDirection: RelationshipDirection.FORWARD,
        sourceEntityKey: '89fac263-2430-48fd-9278-dacfdfc89792',
        targetFilterKeys: [['_type', '_key']],
        targetEntity: {
          _key: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
          _type: 'azure_user_group',
          _class: 'UserGroup',
          displayName: 'Managers',
          jobTitle: null,
          email: null,
        },
      },
      groupId: '89fac263-2430-48fd-9278-dacfdfc89792',
      memberId: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
      memberType: '#microsoft.graph.group',
      displayName: 'HAS',
    });
  });

  test('properties transferred for other', () => {
    expect(
      createGroupMemberRelationship(
        exampleGroupEntity,
        exampleGroupMemberOther,
      ),
    ).toEqual({
      _class: 'HAS',
      _key:
        '89fac263-2430-48fd-9278-dacfdfc89792|has|324e8daa-9c29-42a4-a74b-b9893e6d9750',
      _type: 'azure_group_has_member',
      _mapping: {
        relationshipDirection: RelationshipDirection.FORWARD,
        sourceEntityKey: '89fac263-2430-48fd-9278-dacfdfc89792',
        targetFilterKeys: [['_type', '_key']],
        targetEntity: {
          _key: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
          _type: 'azure_group_member',
          _class: 'User',
          displayName: "Don't really know",
          jobTitle: null,
          email: undefined,
        },
      },
      groupId: '89fac263-2430-48fd-9278-dacfdfc89792',
      memberId: '324e8daa-9c29-42a4-a74b-b9893e6d9750',
      memberType: '#microsoft.graph.directoryObject',
      displayName: 'HAS',
    });
  });
});
