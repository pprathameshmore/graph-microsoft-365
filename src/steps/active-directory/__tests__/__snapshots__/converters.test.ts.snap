// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAccountEntityWithOrganization properties transferred: createAccountEntityWithOrganization 1`] = `
Object {
  "_class": Array [
    "Account",
  ],
  "_key": "microsoft_365_account-the-instance-id",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "intuneConfig": Object {
          "intuneAccountID": "intuneAccountID",
          "mobileDeviceManagementAuthority": "mobileDeviceManagementAuthority",
          "subscriptionState": "subscriptionState",
        },
        "organization": Object {
          "assignedPlans": Array [
            Object {
              "assignedDateTime": "2021-01-13T21:06:18Z",
              "capabilityStatus": "Deleted",
              "service": "SCO",
              "servicePlanId": "1111111-1111-1111-1111-111111111111",
            },
            Object {
              "assignedDateTime": "2021-01-13T21:06:12Z",
              "capabilityStatus": "Enabled",
              "service": "MicrosoftKaizala",
              "servicePlanId": "1111111-1111-1111-1111-111111111111",
            },
          ],
          "businessPhones": Array [],
          "city": undefined,
          "country": undefined,
          "countryLetterCode": "US",
          "createdDateTime": "2019-09-04T16:59:36Z",
          "deletedDateTime": undefined,
          "displayName": "Default Directory",
          "id": "1111111-1111-1111-1111-111111111111",
          "marketingNotificationEmails": Array [],
          "onPremisesLastSyncDateTime": undefined,
          "onPremisesSyncEnabled": undefined,
          "postalCode": undefined,
          "preferredLanguage": "en",
          "privacyProfile": undefined,
          "provisionedPlans": Array [
            Object {
              "capabilityStatus": "Enabled",
              "provisioningStatus": "Success",
              "service": "MicrosoftCommunicationsOnline",
            },
            Object {
              "capabilityStatus": "Enabled",
              "provisioningStatus": "Success",
              "service": "SharePoint",
            },
          ],
          "securityComplianceNotificationMails": Array [],
          "securityComplianceNotificationPhones": Array [],
          "state": undefined,
          "street": undefined,
          "technicalNotificationMails": Array [
            "technicalNotificationMail.jupiterone@hotmail.com",
          ],
          "verifiedDomains": Array [
            Object {
              "capabilities": "Email, OfficeCommunicationsOnline",
              "isDefault": true,
              "isInitial": true,
              "name": "verifiedDomain.onmicrosoft.com",
              "type": "Managed",
            },
          ],
        },
      },
    },
  ],
  "_type": "microsoft_365_account",
  "createdOn": undefined,
  "defaultDomain": "verifiedDomain.onmicrosoft.com",
  "displayName": "instance.config.name configured by customer",
  "id": "1111111-1111-1111-1111-111111111111",
  "intuneAccountId": "intuneAccountID",
  "intuneSubscriptionState": "subscriptionState",
  "mobileDeviceManagementAuthority": "mobileDeviceManagementAuthority",
  "name": "Default Directory",
  "organizationName": "Default Directory",
  "verifiedDomains": Array [
    "verifiedDomain.onmicrosoft.com",
  ],
}
`;
